### PR TITLE
ingress: Add header-case-adjustment enhancement

### PR DESCRIPTION
**Pull request by Miciah**
Proof of concept PRs:

* https://github.com/openshift/api/pull/793
* https://github.com/openshift/router/pull/224
* https://github.com/openshift/cluster-ingress-operator/pull/496

### CODE REVIEW

File: enhancements/ingress/header-case-adjustment.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variable, `ROUTER_H1_CASE_ADJUST`, which specifies adjustments as a
+comma-delimited list of header names in the desired capitalizations.  The router
+parses the value of `ROUTER_H1_CASE_ADJUST` and translates it into
+`h1-case-adjust` settings.  If `ROUTER_H1_CASE_ADJUST` is nonempty, the router
+also specifies the `h1-case-adjust-bogus-server` setting.
```
Comment (sgreene570) [inline enhancements/ingress/header-case-adjustment.md:247]:
```suggestion
also specifies the `h1-case-adjust-bogus-client` setting.
```
The example below shows `h1-case-adjust-bogus-client`, and the example after that shows `h1-case-adjust-bogus-server`, so im assuming you meant `h1-case-adjust-bogus-client` here?

File: enhancements/ingress/header-case-adjustment.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+### Risks and Mitigations
+
+OpenShift cannot infer which HTTP header names legacy clients and servers expect
+to be capitalized a certain way, so cluster administrators and application
+developers need to be aware of the issue and configure IngressControllers and
+Routes using this new configuration when updating to OpenShift 4.7.  A release
+note may mitigate this concern; for example:
+
+"OpenShift 4.7 updates to HAProxy 2.2, which down-cases HTTP header names by
```
Comment (sgreene570) [inline enhancements/ingress/header-case-adjustment.md:287]: OpenShift 4.8?

File: enhancements/ingress/header-case-adjustment.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+default (for example, `Host: xyz.com` is transformed to `host: xyz.com`).  If
+legacy applications are sensitive to the capitalization of HTTP header names,
+see the IngressController's new `spec.httpHeaders.headerNameCaseAdjustments` API
+field for a solution to accommodate these legacy applications until they can be
+fixed."
+
```
Comment (sgreene570) [inline enhancements/ingress/header-case-adjustment.md:293]: One other risk that might be worth including (taken from the[ HAProxy Documentation](https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4-option%20h1-case-adjust-bogus-client)):

```
Applications which fail to properly process requests or responses may require
to temporarily use such workarounds to adjust header names sent to them for
the time it takes the application to be fixed. Please note that an
application which requires such workarounds might be vulnerable to content
smuggling attacks and must absolutely be fixed.
```

I'm not sure if this risk of content smuggling attacks is applicable to our Ingress Controller, so i'll defer to you on whether or not that blurb is worth mentioning (I don't entirely understand how header case adjustments could be a security risk).

File: enhancements/ingress/header-case-adjustment.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+The controller that manages the router Deployment and related resources has unit
+test coverage; for this enhancement, the unit tests have been expanded to cover
+the additional functionality.
+
+The operator has end-to-end tests; for this enhancement, a test has been added
+that (1) creates an IngressController configured with case adjustments, (2)
+creates a Pod and Route for an HTTP application that echoes back requests, (3)
+sends a request to the Route, and (4) verifies that the application echoes back
+the expected headers.
```
Comment (sgreene570) [inline enhancements/ingress/header-case-adjustment.md:306]: nit: the test should also verify that the response itself sets the correct header cases in it's own headers, right?

**Review by sgreene570 — COMMENTED** [review]
Looks great! Some small comments.

File: enhancements/ingress/header-case-adjustment.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+## Alternatives
+
+As an alternative to adding this API, we can add a release note telling cluster
+administrators about the issue and suggesting fixing legacy applications to
+follow the HTTP protocol specification.  Note that OpenShift 4.4, 4.5, and 4.6
+do not exhibit the problem unless HTTP/2 is enabled, which arguably already
+gives users ample time to fix broken applications before upgrading to OpenShift
+4.7.
+
+Instead of introducing this API in OpenShift 4.7, we could revert OpenShift 4.7
```
Comment (sgreene570) [inline enhancements/ingress/header-case-adjustment.md:334]: This alternative is now the chosen approach, right?

File: enhancements/ingress/header-case-adjustment.md (hunk 1 above)
Comment (Miciah) [reply enhancements/ingress/header-case-adjustment.md:247]: Oops!  Well spotted.  I'll fix that.  Thanks!

File: enhancements/ingress/header-case-adjustment.md (hunk 4 above)
Comment (Miciah) [reply enhancements/ingress/header-case-adjustment.md:306]: Right, it checks both.  I'll make that clearer.

File: enhancements/ingress/header-case-adjustment.md (hunk 5 above)
Comment (Miciah) [reply enhancements/ingress/header-case-adjustment.md:334]: Right.  I'll remove that paragraph and revise the preceding one.

File: enhancements/ingress/header-case-adjustment.md [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+Implementing this enhancement required changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variable, `ROUTER_H1_CASE_ADJUST`, which specifies adjustments as a
+comma-delimited list of header names in the desired capitalizations.  The router
```
[the remark this answers is not in the document]
Comment (Miciah) [reply enhancements/ingress/header-case-adjustment.md:244]: I think we are assuming that, and that it is safe to make that assumption.  The cluster administrator can already specify an outrageously long list for `spec.logging.access.httpCaptureHeaders.request`, for example, if the cluster administrator is trying to break the router.

File: enhancements/ingress/header-case-adjustment.md (hunk 3 above)
Comment (Miciah) [reply enhancements/ingress/header-case-adjustment.md:293]: I can't figure out a connection between case adjustment and content smuggling.  I think the point is that an application that is broken in one way (requiring HTTP headers to be capitalized in a certain way) is likely to be broken in another way (not correctly handling chunked encoding).

Comment (sgreene570) [reply enhancements/ingress/header-case-adjustment.md:294]: Fair enough. Thanks for double checking.

**Pull request merged**